include/wally_descriptor.h

```diff
@@ -0,0 +1,172 @@
+ * must be 16 characters or less in length and start with a letter.
+ */
+WALLY_CORE_API int wally_descriptor_parse(
+    const char *descriptor,
+    const struct wally_map *vars_in,
+    uint32_t network,
+    uint32_t flags,
+    struct wally_descriptor **output);
+
+/**
+ * Free a parsed output descriptor or miniscript expression.
+ *
+ * :param descriptor: Parsed output descriptor or miniscript expression to free.
+ */
+WALLY_CORE_API int wally_descriptor_free(
+    struct wally_descriptor *descriptor);
+
+/**
+ * Canonicalize a descriptor.
+ *
+ * :param descriptor: Parsed output descriptor or miniscript expression.
+ * :param flags: For future use. Must be 0.
+ * :param output: Destination for the resulting canonical descriptor.
+ *|    The string returned should be freed using `wally_free_string`.
+ *
+ * .. note:: Other canonicalization (hardened derivation indicator
+ * mapping, and private to public key mapping) is not yet implemented.
+ */
+WALLY_CORE_API int wally_descriptor_canonicalize(
+    const struct wally_descriptor *descriptor,
+    uint32_t flags,
+    char **output);
+
+/**
+ * Create an output descriptor checksum.
+ *
+ * :param descriptor: Parsed output descriptor or miniscript expression.
+ * :param flags: For future use. Must be 0.
+ * :param output: Destination for the resulting descriptor checksum.
+ *|    The string returned should be freed using `wally_free_string`.
+ */
+WALLY_CORE_API int wally_descriptor_get_checksum(
+    const struct wally_descriptor *descriptor,
+    uint32_t flags,
+    char **output);
+
+/**
+ * Get the features used in a parsed output descriptor or miniscript expression.
+ *
+ * :param descriptor: Parsed output descriptor or miniscript expression.
+ * :param value_out: Destination for the resulting ``WALLY_MS_`` feature flags.
+ */
+WALLY_CORE_API int wally_descriptor_get_features(
+    const struct wally_descriptor *descriptor,
+    uint32_t *value_out);
+
+/**
+ * Get the maximum length of a script corresponding to an output descriptor.
+ *
+ * :param descriptor: Parsed output descriptor or miniscript expression.
+ * :param flags: For future use. Must be 0.
+ * :param written: Destination for the resulting maximum script length.
+ *
+ * .. note:: This function overestimates the script size required, but is
+ *|    cheap to compute (does not require script generation).
+ */
+WALLY_CORE_API int wally_descriptor_to_script_get_maximum_length(
+    const struct wally_descriptor *descriptor,
+    uint32_t flags,
+    size_t *written);
+
+/**
+ * Create a script corresponding to an output descriptor or miniscript expression.
+ *
+ * :param descriptor: Parsed output descriptor or miniscript expression.
+ * :param depth: Depth of the expression tree to generate from. Pass 0 to generate from the root.
+ * :param index: The zero-based index of the child at depth ``depth`` to generate from.
+ * :param variant: The variant of descriptor to generate. Pass 0 for the default.
```

Comment:
What is `variant` reserved for?

Reply:
`variant` has been added to support descriptors that return multiple values, the only example of which currently is `combo`. When generating a script/address/address range you normally just have a child number and a bijection between that child number and a given address. The descriptor specs make it clear that to support this bijection, all ranged keys (those with '/*') must be derived at the same child number in a descriptor expression.

 `combo`, along with sub-range key-expressions like rust-bitcoins `<0;1>` extension, break this model and the bijection, which is unfortunate. I consider this a misfeature given (a) there is no specification of the order in which these addresses should be iterated/returned and (b) there is otherwise no mechanism to return them for languages which do no support variant(unitary/list) return types.

Nonetheless we must support it, and `variant` is the mechanism that will allow this. The idea is that each expression construct that breaks the bijection is a variant, and the parsed expression can be queried for the number of variants it contains. variant + child number then biject to a single script/address, with the order documented per-expression type (for `combo` this would be `0=P2PK`, `1=P2PKH`, `2=P2WPKH`, or `3=P2SH-P2WPKH` for example).

Reply:
> we must support it

Do we? `combo` descriptors are only supported by Bitcoin Core for its own legacy wallet backwards compatibility. Each descriptor type has it's own BIP number, so it's possible to simply not support [BIP 384](https://github.com/bitcoin/bips/blob/master/bip-0384.mediawiki). I'm not familiar with rust-bitcoins's `<0;1>` extension.

Reply:
I think we have to support variants, yes. combo is useful for watching/sweeping legacy addresses, and the rust extension suggests there may be a bip44 compatible operator proposed in the future that supports the 0,1 change address delineation when generating addresses (or if not, that ranges themselves will be pushed forward to support this use case).

Reply:
Just a note that this is not that BIP389 descriptors aren't a rust-bitcoin extenstion: https://github.com/bitcoin/bips/pull/1354. I share the sentiment wrt it, but it was a compromise for backups that i now find reasonable.

Reply:
I've pushed full support for `combo()` using variants to demonstrate how this works. I will merge these commits back into the main one for merging.

Reply:
@k-matsuzawa Please note that this changes the behaviour of `combo()` which was previously not quite correct (in that there was no way to generate all variants).

You can now query the number of variants an expression has, and generate scripts/addresses for each variant. Note that for `combo()`, the p2pk variant (0) generates scripts but errors when generating addresses, as p2pk has no address format.

Reply:
> Just a note that BIP389 descriptors (multipath key expressions) aren't a rust-bitcoin extenstion: [bitcoin/bips#1354](https://github.com/bitcoin/bips/pull/1354). I share the sentiment wrt it, but it was a compromise for backups that i now find reasonable.

Thanks, I wasn't aware of this BIP, I've added review comments to it. I think its probably worth adding the range item to the exposed API for when they are eventually implemented in wally.

Reply:
> its probably worth adding the range item to the exposed API

Done.